### PR TITLE
Improve gRPC error message returned when auth fails.

### PR DESCRIPTION
**Pull request by ClifHouck**
Closes https://github.com/G-Research/armada/issues/1370

TODO
- [x] Wrap the rest of errors returned by the anonymous function created in `CreateMiddlewareAuthFunction` with `armadaerrors`.
- [x] Test that errors are actually surfaced with clients of armada.

### CODE REVIEW

File: internal/common/armadaerrors/errors.go [hunk 1]
```diff
@@ -548,3 +590,32 @@ func NewCombinedErrPodUnschedulable(errs ...error) *ErrPodUnschedulable {
 	}
 	return result
 }
+
+// ErrUnauthorized represents an error that occurs when a client tries to
+// perform some action through the gRPC API for which it cannot authenticate.
+//
+// It may be necessary populate the Action field by recovering this error at
+// the gRPC endpoint (using errors.As) and updating the field in-place.
+type ErrUnauthorized struct {
```
Comment (severinson) [inline internal/common/armadaerrors/errors.go:599]: Have you tried this to check the error returned to the user? Please give some examples.

File: internal/common/auth/authorization/common.go [hunk 2]
```diff
@@ -125,15 +128,22 @@ func CreateMiddlewareAuthFunction(authServices []AuthService) grpc_auth.AuthFunc
 	return func(ctx context.Context) (context.Context, error) {
 		for _, service := range authServices {
 			principal, err := service.Authenticate(ctx)
-			if err == missingCredentials {
+			if errors.Is(err, missingCredentials) {
 				// try next auth service
 				continue
+			} else if errors.Is(err, invalidCredentials) {
+				return nil, &armadaerrors.ErrUnauthorized{
```
[the remark this answers is not in the document]
Comment (ClifHouck) [reply internal/common/auth/authorization/common.go]: Addressed in b5dc1d12feafff71ae6bbc369ce39fac1f276550 and 259bf19049d9b6b09318ffbb4d1655baa0d0faaf

File: internal/common/armadaerrors/errors.go [hunk 3]
```diff
@@ -278,6 +278,12 @@ func CodeFromError(err error) codes.Code {
 			return codes.InvalidArgument
 		}
 	}
+	{
+		var e *ErrUnauthorized
```
[the remark this answers is not in the document]
Comment (ClifHouck) [reply internal/common/armadaerrors/errors.go]: Addressed in 259bf19049d9b6b09318ffbb4d1655baa0d0faaf

File: internal/common/armadaerrors/errors.go (hunk 1 above)
Comment (ClifHouck) [reply internal/common/armadaerrors/errors.go]: `2022/10/04 14:10:52 error submitting request &api.JobSubmitRequest{Queue:"test", JobSetId:"job-set-1", JobRequestItems:[]*api.JobSubmitRequestItem{(*api.JobSubmitRequestItem)(0xc0005b00a0)}}: rpc error: code = Unauthenticated desc = [x-request-id: "JLdcWD3VE5CmFqUoCzpfz"] Could not authorize user "<none>" via service "Basic" while attempting action "/api.Submit/SubmitJobs"; rpc error: code = Unauthenticated desc = invalid username/password` Looks good from `armadactl`'s view.

**Comment by ClifHouck** [comment]
This needs a fresh mind in the morning, but I think it's getting closer to where it should be.

File: internal/common/armadaerrors/errors.go [hunk 4]
```diff
@@ -548,3 +584,130 @@ func NewCombinedErrPodUnschedulable(errs ...error) *ErrPodUnschedulable {
...
+func (err *ErrInternalAuthServiceError) GRPCStatus() *status.Status {
+	// TODO(clif) Whats the right code to return here or should we give the
+	// auth service the opportunity to set it?
+	return status.New(codes.Unavailable, err.Error())
+}
+
+// ErrInternalAuthServiceError is returned when an auth service encounters
+// an internal error that is not directly related to the supplied input/
+// credentials.
+func (err *ErrInternalAuthServiceError) Error() string {
```
Comment (severinson) [inline internal/common/armadaerrors/errors.go:684]: I'd move this docstring to the error type definition.

File: internal/common/auth/authorization/basic_test.go [hunk 5]
```diff
@@ -26,10 +28,13 @@ func TestBasicAuthService(t *testing.T) {
 		metadata.NewIncomingContext(context.Background(), basicPassword("root", "test")))
 
 	assert.NotNil(t, e)
-	assert.NotEqual(t, e, missingCredentials)
+	fmt.Println(e.Error())
```
Comment (severinson) [inline internal/common/auth/authorization/basic_test.go:31]: Remove println

Comment (ClifHouck) [reply internal/common/auth/authorization/basic_test.go]: Addressed in e639556

**Comment by ClifHouck** [comment]
I think this is mostly GTG.

I've made a companion PR to unit test `KerberosAuthService.Authenticate` here: https://github.com/ClifHouck/armada/pull/1/files, which I'll re-open/re-pull against G-Research/armada once this PR merges.

File: internal/common/armadaerrors/errors.go [hunk 6]
```diff
@@ -385,6 +385,38 @@ func IsNetworkError(err error) bool {
 	return false
 }
 
+// Add the action to th error if possible.
```
Comment (richscott) [inline internal/common/armadaerrors/errors.go:388]: Should `th` be `the` ?

File: internal/common/armadaerrors/errors.go [hunk 7]
```diff
@@ -385,6 +385,38 @@ func IsNetworkError(err error) bool {
...
+	}
+	{
+		var e *ErrUnauthorized
+		if errors.As(err, &e) {
+			e.Action = info.FullMethod
+		}
+	}
+}
+
+// Add the action to th error if possible.
```
Comment (richscott) [inline internal/common/armadaerrors/errors.go:404]: Here as well?

File: internal/common/auth/authorization/kubernetes.go [hunk 8]
```diff
@@ -161,12 +183,20 @@ func (authService *KubernetesNativeAuthService) getClusterURL(token string) (str
 func (authService *KubernetesNativeAuthService) reviewToken(ctx context.Context, clusterUrl string, token string, ca []byte) (string, error) {
 	result, err := authService.TokenReviewer.ReviewToken(ctx, clusterUrl, token, ca)
 	if err != nil {
-		return "", err
+		// TODO(clif) Hard to tell if this should be internal auth error
+		// or invalid creds still.
```
Comment (richscott) [inline internal/common/auth/authorization/kubernetes.go:187]: Does ReviewToken() return different errors that delineate if it's a mangled/bad token as opposed to the auth server erroring?

**Review by richscott — COMMENTED** [review]
Just a couple typos in comments, and a question about an error return, otherwise it looks good.

File: internal/common/auth/authorization/kubernetes.go (hunk 8 above)
Comment (ClifHouck) [reply internal/common/auth/authorization/kubernetes.go:187]: Really hard to tell, as the call just seems to end up doing a `POST` to this kubernetes endpoint: https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-review-v1/ Doesn't seem to delineate what errors could be returned.

File: internal/common/armadaerrors/errors.go (hunk 6 above)
Comment (ClifHouck) [reply internal/common/armadaerrors/errors.go]: Addressed in 1de8695

**Review by richscott — APPROVED** [review]

**Pull request merged**